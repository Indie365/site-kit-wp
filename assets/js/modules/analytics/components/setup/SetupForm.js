/**
 * Analytics Setup form.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { useCallback, useEffect } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import Data from 'googlesitekit-data';
import { Button, ProgressBar } from 'googlesitekit-components';
import {
	SETUP_FLOW_MODE_UA,
	SETUP_FLOW_MODE_GA4,
	MODULES_ANALYTICS,
	FORM_SETUP,
	EDIT_SCOPE,
} from '../../datastore/constants';
import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
import { CORE_FORMS } from '../../../../googlesitekit/datastore/forms/constants';
import { isPermissionScopeError } from '../../../../util/errors';
import SetupFormUA from './SetupFormUA';
import SetupFormGA4 from './SetupFormGA4';
import StoreErrorNotices from '../../../../components/StoreErrorNotices';
import { MODULES_ANALYTICS_4 } from '../../../analytics-4/datastore/constants';
const { useSelect, useDispatch } = Data;

export default function SetupForm( { finishSetup } ) {
	const canSubmitUAChanges = useSelect( ( select ) =>
		select( MODULES_ANALYTICS ).canSubmitChanges()
	);
	const hasEditScope = useSelect( ( select ) =>
		select( CORE_USER ).hasScope( EDIT_SCOPE )
	);
	const autoSubmit = useSelect( ( select ) =>
		select( CORE_FORMS ).getValue( FORM_SETUP, 'autoSubmit' )
	);
	const setupFlowMode = useSelect( ( select ) =>
		select( MODULES_ANALYTICS ).getSetupFlowMode()
	);
	const isUAEnabled = useSelect( ( select ) =>
		select( CORE_FORMS ).getValue( FORM_SETUP, 'enableUA' )
	);
	const canSubmitGA4Changes = useSelect( ( select ) => {
		const canSubmitChanges =
			select( MODULES_ANALYTICS_4 ).canSubmitChanges();

		if ( isUAEnabled ) {
			return canSubmitUAChanges && canSubmitChanges;
		}

		return canSubmitChanges;
	} );

	const { setValues } = useDispatch( CORE_FORMS );
	const { submitChanges } = useDispatch( MODULES_ANALYTICS );

	const submitForm = useCallback(
		async ( event ) => {
			event.preventDefault();
			// Disable autoSubmit unconditionally to prevent
			// automatic invocation more than once.
			setValues( FORM_SETUP, { autoSubmit: false } );

			const { error } = await submitChanges();

			if ( isPermissionScopeError( error ) ) {
				setValues( FORM_SETUP, { autoSubmit: true } );
			}

			if ( ! error ) {
				finishSetup();
			}
		},
		[ finishSetup, setValues, submitChanges ]
	);

	const gtmContainersResolved = useSelect( ( select ) =>
		select( MODULES_ANALYTICS ).hasFinishedLoadingGTMContainers()
	);

	// If the user lands back on this component with autoSubmit and the edit scope,
	// resubmit the form.
	useEffect( () => {
		if ( autoSubmit && hasEditScope ) {
			submitForm( { preventDefault: () => {} } );
		}
	}, [ hasEditScope, autoSubmit, submitForm ] );

	if ( ! gtmContainersResolved ) {
		return <ProgressBar />;
	}

	return (
		<form
			className="googlesitekit-analytics-setup__form"
			onSubmit={ submitForm }
		>
			<StoreErrorNotices
				moduleSlug="analytics"
				storeName={ MODULES_ANALYTICS }
			/>
			<StoreErrorNotices
				moduleSlug="analytics-4"
				storeName={ MODULES_ANALYTICS_4 }
			/>
			{ setupFlowMode === SETUP_FLOW_MODE_UA && <SetupFormUA /> }
			{ setupFlowMode === SETUP_FLOW_MODE_GA4 && <SetupFormGA4 /> }
			<div className="googlesitekit-setup-module__action">
				<Button disabled={ ! canSubmitGA4Changes }>
					{ __( 'Configure Analytics', 'google-site-kit' ) }
				</Button>
			</div>
		</form>
	);
}

SetupForm.propTypes = {
	finishSetup: PropTypes.func,
};

SetupForm.defaultProps = {
	finishSetup: () => {},
};
