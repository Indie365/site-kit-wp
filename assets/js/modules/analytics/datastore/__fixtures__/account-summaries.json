[
  {
    "id": "22334455",
    "kind": "analytics#accountSummary",
    "name": "Account A",
    "starred": null,
    "webProperties": [
      {
        "id": "UA-22334455-1",
        "internalWebPropertyId": "123456789",
        "kind": "analytics#webPropertySummary",
        "level": "STANDARD",
        "name": "Account A - Property A",
        "starred": null,
        "websiteUrl": "http://example.com",
        "profiles": [
          {
            "id": "123456789",
            "kind": "analytics#profileSummary",
            "name": "All Web Site Data",
            "starred": null,
            "type": "WEB"
          }
        ]
      }
    ]
  },
  {
    "id": "33334444",
    "kind": "analytics#accountSummary",
    "name": "Account B",
    "starred": null,
    "webProperties": [
    ]
  }
]
