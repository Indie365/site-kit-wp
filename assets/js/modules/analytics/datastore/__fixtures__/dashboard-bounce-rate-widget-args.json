{
	"dateRange": "last-28-days",
	"multiDateRange": 1,
	"dimensions": "ga:date",
	"metrics": [
		{
			"expression": "ga:bounceRate",
			"alias": "Bounce Rate"
		}
	]
}