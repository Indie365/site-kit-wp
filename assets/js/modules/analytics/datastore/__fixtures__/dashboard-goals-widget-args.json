{
	"dateRange": "last-28-days",
	"multiDateRange": 1,
	"dimensions": "ga:date",
	"metrics": [
		{
			"expression": "ga:goalCompletionsAll",
			"alias": "Goal Completions"
		}
	]
}