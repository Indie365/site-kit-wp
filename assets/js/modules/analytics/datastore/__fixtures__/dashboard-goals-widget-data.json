[
	{
		"nextPageToken": null,
		"columnHeader": {
			"dimensions": [
				"ga:date"
			],
			"metricHeader": {
				"metricHeaderEntries": [
					{
						"name": "Goal Completions",
						"type": "INTEGER"
					}
				]
			}
		},
		"data": {
			"dataLastRefreshed": null,
			"isDataGolden": null,
			"rowCount": null,
			"samplesReadCounts": null,
			"samplingSpaceSizes": null,
			"rows": [
				{
					"dimensions": [
						"20200813"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200814"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200815"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200816"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200817"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200818"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200819"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200820"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200821"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200822"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200823"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200824"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200825"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200826"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200827"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200828"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200829"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200830"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200831"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200901"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200902"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200903"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200904"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200905"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200906"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200907"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200908"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				},
				{
					"dimensions": [
						"20200909"
					],
					"metrics": [
						{
							"values": [
								"0"
							]
						}
					]
				}
			],
			"totals": [
				{
					"values": [
						"0"
					]
				},
				{
					"values": [
						"0"
					]
				}
			]
		}
	}
]