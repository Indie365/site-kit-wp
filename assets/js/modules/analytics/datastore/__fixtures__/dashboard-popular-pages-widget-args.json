{
	"dateRange": "last-28-days",
	"dimensions": "ga:pageTitle,ga:pagePath",
	"metrics": [
		{
			"expression": "ga:pageviews",
			"alias": "Pageviews"
		}
	],
	"orderby": [
		{
			"fieldName": "ga:pageviews",
			"sortOrder": "DESCENDING"
		}
	],
	"limit": 10
}