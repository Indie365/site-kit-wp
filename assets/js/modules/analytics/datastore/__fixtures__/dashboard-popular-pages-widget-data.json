[
	{
		"nextPageToken": "10",
		"columnHeader": {
			"dimensions": [
				"ga:pageTitle",
				"ga:pagePath"
			],
			"metricHeader": {
				"metricHeaderEntries": [
					{
						"name": "Pageviews",
						"type": "INTEGER"
					}
				]
			}
		},
		"data": {
			"dataLastRefreshed": null,
			"isDataGolden": null,
			"rowCount": 177,
			"samplesReadCounts": null,
			"samplingSpaceSizes": null,
			"rows": [
				{
					"dimensions": [
						"Site Kit Test 1",
						"/"
					],
					"metrics": [
						{
							"values": [
								"1469"
							]
						}
					]
				},
				{
					"dimensions": [
						"Site Kit Test 2",
						"/site-kit-tests/"
					],
					"metrics": [
						{
							"values": [
								"616"
							]
						}
					]
				},
				{
					"dimensions": [
						"Site Kit Test 3",
						"/site-kit-tests/"
					],
					"metrics": [
						{
							"values": [
								"302"
							]
						}
					]
				},
				{
					"dimensions": [
						"Site Kit Test 4",
						"/site-kit-tests/"
					],
					"metrics": [
						{
							"values": [
								"231"
							]
						}
					]
				},
				{
					"dimensions": [
						"Site Kit Tests 5",
						"/site-kit-tests/"
					],
					"metrics": [
						{
							"values": [
								"203"
							]
						}
					]
				},
				{
					"dimensions": [
						"Site Kit Tests 6",
						"/site-kit-tests/"
					],
					"metrics": [
						{
							"values": [
								"110"
							]
						}
					]
				},
				{
					"dimensions": [
						"Site Kit Tests 7",
						"/site-kit-tests/"
					],
					"metrics": [
						{
							"values": [
								"98"
							]
						}
					]
				},
				{
					"dimensions": [
						"Site Kit Tests 8",
						"/site-kit-tests/"
					],
					"metrics": [
						{
							"values": [
								"78"
							]
						}
					]
				},
				{
					"dimensions": [
						"Site Kit Tests 9",
						"/site-kit-tests/"
					],
					"metrics": [
						{
							"values": [
								"72"
							]
						}
					]
				},
				{
					"dimensions": [
						"Site Kit Tests 10",
						"/site-kit-tests/"
					],
					"metrics": [
						{
							"values": [
								"71"
							]
						}
					]
				}
			],
			"totals": [
				{
					"values": [
						"4264"
					]
				}
			],
			"minimums": [
				{
					"values": [
						"1"
					]
				}
			],
			"maximums": [
				{
					"values": [
						"1469"
					]
				}
			]
		}
	}
]