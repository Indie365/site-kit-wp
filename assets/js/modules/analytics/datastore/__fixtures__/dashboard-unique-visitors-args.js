export const dashboardUniqueVisitorsVisitorArgs = {
	dateRange: 'last-28-days',
	multiDateRange: 1,
	metrics: [
		{
			expression: 'ga:users',
			alias: 'Total Users',
		},
	],
};

export const dashboardUniqueVisitorsSparkArgs = {
	dateRange: 'last-28-days',
	dimensions: 'ga:date',
	metrics: [
		{
			expression: 'ga:users',
			alias: 'Users',
		},
	],

};
