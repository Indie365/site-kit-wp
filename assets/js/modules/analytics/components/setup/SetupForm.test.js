/**
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { act, fireEvent, render } from '../../../../../../tests/js/test-utils';
import {
	createTestRegistry,
	provideModules,
	provideSiteInfo,
	provideUserAuthentication,
	waitForDefaultTimeouts,
} from '../../../../../../tests/js/utils';
import { CORE_FORMS } from '../../../../googlesitekit/datastore/forms/constants';
import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
import { MODULES_ANALYTICS_4 } from '../../../analytics-4/datastore/constants';
import { MODULES_TAGMANAGER } from '../../../tagmanager/datastore/constants';
import {
	EDIT_SCOPE,
	FORM_SETUP,
	MODULES_ANALYTICS,
	SETUP_FLOW_MODE_GA4,
} from '../../datastore/constants';
import * as fixtures from '../../../analytics-4/datastore/__fixtures__';
import ga4ReportingTour from '../../../../feature-tours/ga4-reporting';
import SetupForm from './SetupForm';
import { ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY } from '../../../analytics-4/constants';

const {
	accountSummaries,
	defaultEnhancedMeasurementSettings,
	webDataStreamsBatch,
	webDataStreams,
} = fixtures;
const accounts = accountSummaries;
const properties = accounts[ 1 ].propertySummaries;
const accountID = accounts[ 1 ]._id;
const propertyID = properties[ 0 ]._id;
const webDataStreamID = webDataStreamsBatch[ propertyID ][ 0 ]._id;

describe( 'SetupForm', () => {
	let registry;

	beforeEach( () => {
		registry = createTestRegistry();
		provideSiteInfo( registry );
		provideUserAuthentication( registry );
		provideModules( registry, [ { slug: 'analytics', active: true } ] );
	} );

	it( 'renders the form with a progress bar when GTM containers are not resolved', () => {
		provideModules( registry, [
			{ slug: 'analytics', active: true, connected: true },
			{ slug: 'tagmanager', active: true, connected: true },
		] );
		registry.dispatch( MODULES_ANALYTICS ).setSettings( {} );
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {} );
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );

		const { container, getByRole } = render( <SetupForm />, { registry } );

		expect( container ).toMatchSnapshot();

		expect( getByRole( 'progressbar' ) ).toBeInTheDocument();
	} );

	it( 'renders the form correctly when setup flow mode is GA4', async () => {
		registry.dispatch( MODULES_ANALYTICS ).setSettings( {} );
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {} );
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetAccountSummaries( accountSummaries );
		registry
			.dispatch( MODULES_ANALYTICS )
			.receiveGetProperties( [], { accountID } );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetProperty( properties[ 0 ], {
				propertyID,
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
				propertyIDs: [ propertyID ],
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetEnhancedMeasurementSettings(
				{
					...defaultEnhancedMeasurementSettings,
					streamEnabled: false,
				},
				{
					propertyID,
					webDataStreamID,
				}
			);

		await registry.dispatch( MODULES_ANALYTICS ).selectAccount( accountID );

		// Verify that the setup flow mode is GA4.
		expect( registry.select( MODULES_ANALYTICS ).getSetupFlowMode() ).toBe(
			SETUP_FLOW_MODE_GA4
		);

		const { container, getByText, waitForRegistry } = render(
			<SetupForm />,
			{
				registry,
			}
		);
		await waitForRegistry();

		expect( container ).toMatchSnapshot();

		expect( getByText( 'Account' ) ).toBeInTheDocument();
		expect( getByText( 'Property' ) ).toBeInTheDocument();
		expect( getByText( 'Web Data Stream' ) ).toBeInTheDocument();
	} );

	it( 'submits the form upon pressing the CTA', async () => {
		registry.dispatch( MODULES_ANALYTICS ).setSettings( {} );
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {} );
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetAccountSummaries( accountSummaries );
		registry
			.dispatch( MODULES_ANALYTICS )
			.receiveGetProperties( [], { accountID } );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetProperty( properties[ 0 ], {
				propertyID,
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
				propertyIDs: [ propertyID ],
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetEnhancedMeasurementSettings(
				defaultEnhancedMeasurementSettings,
				{
					propertyID,
					webDataStreamID,
				}
			);

		await registry.dispatch( MODULES_ANALYTICS ).selectAccount( accountID );

		const finishSetup = jest.fn();
		const { getByRole, waitForRegistry } = render(
			<SetupForm finishSetup={ finishSetup } />,
			{
				registry,
			}
		);
		await waitForRegistry();

		act( () => {
			// It doesn't seem possible to show the dropdown menu within the test by clicking on the dropdown in the usual way.
			// Therefore, we simulate the click on the dropdown menu item directly, despite it being hidden.
			fireEvent.click(
				getByRole( 'menuitem', {
					name: /Example Property/i,
					hidden: true,
				} )
			);
		} );

		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedTours( [ ga4ReportingTour.slug ] );

		const updateAnalyticsSettingsRegexp = new RegExp(
			'/analytics/data/settings'
		);

		const updateAnalytics4SettingsRegexp = new RegExp(
			'/analytics-4/data/settings'
		);

		const dismissItemEndpointRegexp = new RegExp(
			'^/google-site-kit/v1/core/user/data/dismiss-item'
		);

		fetchMock.post( updateAnalyticsSettingsRegexp, {
			status: 200,
			body: {},
		} );

		fetchMock.post( updateAnalytics4SettingsRegexp, {
			status: 200,
			body: {},
		} );

		fetchMock.post( dismissItemEndpointRegexp, {
			status: 200,
			body: JSON.stringify( [
				ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY,
			] ),
		} );

		act( () => {
			fireEvent.click(
				getByRole( 'button', { name: /Configure Analytics/i } )
			);
		} );

		await waitForRegistry();

		// An additional wait is required in order for all resolvers to finish.
		await act( waitForDefaultTimeouts );

		expect( fetchMock ).toHaveFetchedTimes(
			1,
			updateAnalyticsSettingsRegexp
		);
		expect( fetchMock ).toHaveFetchedTimes(
			1,
			updateAnalytics4SettingsRegexp
		);

		expect( finishSetup ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'auto-submits the form', async () => {
		registry.dispatch( MODULES_ANALYTICS ).setSettings( {} );
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
		registry
			.dispatch( MODULES_ANALYTICS )
			.receiveGetProperties( [], { accountID } );
		registry.dispatch( MODULES_ANALYTICS ).receiveGetExistingTag( null );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetExistingTag( null );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetAccountSummaries(
			accountSummaries.map( ( account ) => ( {
				...account,
				propertySummaries: [],
			} ) )
		);
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetProperty( properties[ 0 ], {
				propertyID,
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
				propertyIDs: [ propertyID ],
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetEnhancedMeasurementSettings(
				defaultEnhancedMeasurementSettings,
				{
					propertyID,
					webDataStreamID,
				}
			);

		await registry.dispatch( MODULES_ANALYTICS ).selectAccount( accountID );

		// Simulate an auto-submit case where the user is returning to the page
		// after granting extra scopes necessary to submit.
		// In this situation, the autoSubmit is set before the user goes to oAuth,
		// store state is snapshotted, and then restored upon returning.
		registry
			.dispatch( CORE_FORMS )
			.setValues( FORM_SETUP, { autoSubmit: true } );

		provideUserAuthentication( registry, {
			grantedScopes: [ EDIT_SCOPE ],
		} );

		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedTours( [ ga4ReportingTour.slug ] );

		const createPropertyRegexp = new RegExp(
			'/analytics-4/data/create-property'
		);

		const createWebDataStreamRegexp = new RegExp(
			'/analytics-4/data/create-webdatastream'
		);

		const updateAnalyticsSettingsRegexp = new RegExp(
			'/analytics/data/settings'
		);

		const updateAnalytics4SettingsRegexp = new RegExp(
			'/analytics-4/data/settings'
		);

		fetchMock.post( createPropertyRegexp, {
			status: 200,
			body: properties[ 0 ],
		} );

		fetchMock.post( createWebDataStreamRegexp, {
			status: 200,
			body: webDataStreams[ 2 ],
		} );

		fetchMock.post( updateAnalyticsSettingsRegexp, {
			status: 200,
			body: {},
		} );

		fetchMock.post( updateAnalytics4SettingsRegexp, {
			status: 200,
			body: {},
		} );

		const finishSetup = jest.fn();
		const { getByRole, waitForRegistry } = render(
			<SetupForm finishSetup={ finishSetup } />,
			{
				registry,
			}
		);
		await waitForRegistry();

		// Ensure the form rendered successfully.
		expect(
			getByRole( 'button', { name: /Configure Analytics/i } )
		).toBeInTheDocument();

		await waitForRegistry();

		// An additional wait is required in order for all resolvers to finish.
		await act( async () => {
			await waitForDefaultTimeouts();
		} );

		expect( fetchMock ).toHaveFetchedTimes( 1, createPropertyRegexp );
		expect( fetchMock ).toHaveFetchedTimes( 1, createWebDataStreamRegexp );
		expect( fetchMock ).toHaveFetchedTimes(
			1,
			updateAnalyticsSettingsRegexp
		);
		expect( fetchMock ).toHaveFetchedTimes(
			1,
			updateAnalytics4SettingsRegexp
		);

		expect( finishSetup ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'auto-submits the form only once in the case of an error', async () => {
		registry.dispatch( MODULES_ANALYTICS ).setSettings( {} );
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
		registry
			.dispatch( MODULES_ANALYTICS )
			.receiveGetProperties( [], { accountID } );
		registry.dispatch( MODULES_ANALYTICS ).receiveGetExistingTag( null );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetExistingTag( null );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetAccountSummaries(
			accountSummaries.map( ( account ) => ( {
				...account,
				propertySummaries: [],
			} ) )
		);
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
				propertyIDs: [ propertyID ],
			} );
		await registry.dispatch( MODULES_ANALYTICS ).selectAccount( accountID );

		// Simulate an auto-submit case where the user is returning to the page
		// after granting extra scopes necessary to submit.
		// In this situation, the autoSubmit is set before the user goes to oAuth,
		// store state is snapshotted, and then restored upon returning.
		registry
			.dispatch( CORE_FORMS )
			.setValues( FORM_SETUP, { autoSubmit: true } );
		provideUserAuthentication( registry, {
			grantedScopes: [ EDIT_SCOPE ],
		} );

		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedTours( [ ga4ReportingTour.slug ] );

		const createPropertyRegexp = new RegExp(
			'/analytics-4/data/create-property'
		);

		fetchMock.post( createPropertyRegexp, {
			status: 403,
			body: {
				code: 403,
				error: 'Insufficient permissions',
			},
		} );

		const finishSetup = jest.fn();
		const { getByRole, waitForRegistry } = render(
			<SetupForm finishSetup={ finishSetup } />,
			{
				registry,
			}
		);
		await waitForRegistry();

		// Ensure the form rendered successfully.
		expect(
			getByRole( 'button', { name: /Configure Analytics/i } )
		).toBeInTheDocument();

		// While not strictly needed, add waits to match the successful auto-submit test case to help avoid a false positive result.
		await waitForRegistry();
		await act( async () => {
			await waitForDefaultTimeouts();
		} );

		// Create property should have only been called once.
		expect( fetchMock ).toHaveFetchedTimes( 1, createPropertyRegexp );
		// Setup was not successful, so the finish function should not be called.
		expect( finishSetup ).not.toHaveBeenCalled();
		// Expect a console error due to the API error (otherwise this test will fail).
		expect( console ).toHaveErrored();
	} );
} );
