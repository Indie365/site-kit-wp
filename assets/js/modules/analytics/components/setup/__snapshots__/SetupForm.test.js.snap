// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetupForm renders the form correctly when setup flow mode is GA4 1`] = `
<div>
  <form
    class="googlesitekit-analytics-setup__form"
  >
    <p
      class="googlesitekit-margin-bottom-0"
    >
      Please select the account information below. You can change this later in your settings.
    </p>
    <div
      class="googlesitekit-setup-module__inputs"
    >
      <div
        class="mdc-select googlesitekit-analytics__select-account mdc-select--outlined"
      >
        <i
          class="mdc-select__dropdown-icon"
        />
        <input
          name="enhanced-select"
          type="hidden"
          value="22305160"
        />
        <div
          aria-disabled="false"
          class="mdc-select__selected-text"
          tabindex="0"
        >
          10up
        </div>
        <div
          class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
        >
          <div
            class="mdc-notched-outline__leading"
          />
          <div
            class="mdc-notched-outline__notch"
          >
            <label
              class="mdc-floating-label mdc-floating-label--float-above"
              for="googlesitekit-select-0"
              style=""
            >
              Account
            </label>
          </div>
          <div
            class="mdc-notched-outline__trailing"
          />
        </div>
      </div>
      <div
        class="mdc-select googlesitekit-analytics-4__select-property googlesitekit-analytics-4__select-property--loaded mdc-select--outlined"
      >
        <i
          class="mdc-select__dropdown-icon"
        />
        <input
          name="enhanced-select"
          type="hidden"
          value="property_create"
        />
        <div
          aria-disabled="false"
          class="mdc-select__selected-text"
          tabindex="0"
        >
          Set up a new property
        </div>
        <div
          class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
        >
          <div
            class="mdc-notched-outline__leading"
          />
          <div
            class="mdc-notched-outline__notch"
          >
            <label
              class="mdc-floating-label mdc-floating-label--float-above"
              for="googlesitekit-select-1"
              style=""
            >
              Property
            </label>
          </div>
          <div
            class="mdc-notched-outline__trailing"
          />
        </div>
      </div>
      <div
        class="mdc-select googlesitekit-analytics-4__select-webdatastream mdc-select--outlined"
      >
        <i
          class="mdc-select__dropdown-icon"
        />
        <input
          name="enhanced-select"
          type="hidden"
          value="webdatastream_create"
        />
        <div
          aria-disabled="false"
          class="mdc-select__selected-text"
          tabindex="0"
        >
          Set up a new web data stream
        </div>
        <div
          class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
        >
          <div
            class="mdc-notched-outline__leading"
          />
          <div
            class="mdc-notched-outline__notch"
          >
            <label
              class="mdc-floating-label mdc-floating-label--float-above"
              for="googlesitekit-select-2"
              style=""
            >
              Web Data Stream
            </label>
          </div>
          <div
            class="mdc-notched-outline__trailing"
          />
        </div>
      </div>
    </div>
    <div
      class="googlesitekit-setup-module__action"
    >
      <button
        class="mdc-button mdc-button--raised"
        target="_self"
      >
        <span
          class="mdc-button__label"
        >
          Configure Analytics
        </span>
      </button>
    </div>
  </form>
</div>
`;

exports[`SetupForm renders the form with a progress bar when GTM containers are not resolved 1`] = `
<div>
  <div
    class="mdc-linear-progress mdc-linear-progress--indeterminate"
    role="progressbar"
  >
    <div
      class="mdc-linear-progress__buffering-dots"
    />
    <div
      class="mdc-linear-progress__buffer"
    />
    <div
      class="mdc-linear-progress__bar mdc-linear-progress__primary-bar"
    >
      <span
        class="mdc-linear-progress__bar-inner"
      />
    </div>
    <div
      class="mdc-linear-progress__bar mdc-linear-progress__secondary-bar"
    >
      <span
        class="mdc-linear-progress__bar-inner"
      />
    </div>
  </div>
</div>
`;
